### PR TITLE
feat: Update keystore structure

### PR DESCRIPTION
**Pull request by seebees**
Implement the specification changes to update the keystore structure.
Also adds input to the CreateKey operation
to allow creating a specific branch key id
as well as adding custom encryption context to a branch key.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStore/src/KMSKeystoreOperations.dfy [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+    var reEncryptResponse :- maybeReEncryptResponse
+    .MapFailure(e => Types.ComAmazonawsKms(ComAmazonawsKms := e));
+
+    :- Need(
+      && reEncryptResponse.SourceKeyId.Some?
+      && reEncryptResponse.KeyId.Some?
+      && reEncryptResponse.SourceKeyId.value == kmsConfiguration.kmsKeyArn
+      && reEncryptResponse.KeyId.value == kmsConfiguration.kmsKeyArn,
+      // && ParseAwsKmsIdentifier(reEncryptResponse.SourceKeyId.value).Success?
+      // && ParseAwsKmsIdentifier(reEncryptResponse.KeyId.value).Success?,
```
Comment (josecorella) [inline AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStore/src/KMSKeystoreOperations.dfy:157]:
```suggestion
```

**Review by josecorella — CHANGES_REQUESTED** [review]

File: AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStore/src/DDBKeystoreOperations.dfy [hunk 2]
```diff
@@ -0,0 +1,374 @@
...
+    ensures output.Success?
+            ==>
+              && output.value[Structure.BRANCH_KEY_IDENTIFIER_FIELD].S == branchKeyIdentifier
+              && Seq.Last(ddbClient.History.GetItem).output.Success?
+              && Seq.Last(ddbClient.History.GetItem).output.value.Item.Some?
+              && output == Success(Seq.Last(ddbClient.History.GetItem).output.value.Item.value)
+
+    ensures
+      && old(ddbClient.History.GetItem) < ddbClient.History.GetItem
+      && old(ddbClient.History.TransactWriteItems) == ddbClient.History.TransactWriteItems
```
[the remark this answers is not in the document]
Comment (seebees) [reply AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStore/src/DDBKeystoreOperations.dfy:170]: Yes, we do need it.
This is a specification that says we are not doing `TransactWriteItems`.
Without this, Dafny would not believe that `TransactWriteItems` History is unchanged.

File: AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStore/src/GetKeys.dfy [hunk 3]
```diff
@@ -2,441 +2,507 @@
...
-              && output.value.KeyId.Some?
-                 //= aws-encryption-sdk-specification/framework/branch-key-store.md#getbeaconkey
-                 //= type=implication
-                 //# - The `KeyId` field in the AWS KMS response MUST equal the configured AWS KMS Key ARN.
-              && output.value.KeyId.value == awsKmsKeyArn
-              && output.value.Plaintext.Some?
-              && 32 == |output.value.Plaintext.value|
-  {
-    var wrappedBranchKey: KMS.CiphertextType := branchKeyRecord[BRANCH_KEY_FIELD].B;
+    // && |kmsClient.History.Decrypt| == |old(kmsClient.History.Decrypt)| + 1
```
[the remark this answers is not in the document]
Comment (seebees) [reply AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStore/src/GetKeys.dfy:465]: Moved to a better place

**Review by josecorella — APPROVED** [review]
lgtm

**Pull request merged**